proof/invariant-abstract/X64/ArchInvariants_AI.thy

```diff
@@ -1734,8 +1753,51 @@ lemma global_refs_update [iff]:
 
 end
 
+context Arch_p_asid_table_update_eq begin
+
+lemma vs_lookup_update[iff]:
+  "vs_lookup (f s) = vs_lookup s"
+  by (auto elim!: vs_lookup_stateI simp: asid_table obj_at_def)
+
+lemma vspace_at_asid_update[iff]:
+  "vspace_at_asid asid pt (f s) = vspace_at_asid asid pt s"
+  by (simp add: vspace_at_asid_def)
+
+lemma vs_lookup_pages_update[iff]:
+  "vs_lookup_pages (f s) = vs_lookup_pages s"
+  by (auto elim!: vs_lookup_pages_stateI simp: asid_table obj_at_def)
+
+lemma valid_vs_lookup_update [iff]:
+  "valid_vs_lookup (f s) = valid_vs_lookup s"
+  by (simp add: valid_vs_lookup_def asid_table )
+
+lemma valid_table_caps_update [iff]:
+  "valid_table_caps (f s) = valid_table_caps s"
+  by (simp add: valid_table_caps_def asid_table pspace second_level_tables)
+
+lemma valid_vspace_objs_update [iff]:
+  "valid_vspace_objs (f s) = valid_vspace_objs s"
+  by (simp add: valid_vspace_objs_def pspace)
+
+lemma valid_arch_caps_update [iff]:
+  "valid_arch_caps (f s) = valid_arch_caps s"
+  by (simp add: valid_arch_caps_def asid_table)
+
+lemma valid_kernel_mappings_update [iff]:
+  "valid_kernel_mappings (f s) = valid_kernel_mappings s"
+  by (simp add: valid_kernel_mappings_def pspace second_level_tables)
+
+lemma valid_asid_map_update [iff]:
+  "valid_asid_map (f s) = valid_asid_map s"
+  by (simp add: valid_asid_map_def)
+
+end
+
 context Arch_p_arch_update_eq begin
 
+sublocale Arch_p_asid_table_update_eq
+  by unfold_locales (simp add: arch second_level_tables_def)+
```

Comment:
still not following what's happening here... for some reason I'm under the impression that lemmas in this locale that do not depend on the inclusion of Arch_p_asid_table_update_eq should go above the sublocale command, but here we have it at the top of the context block

Reply:
I have to admit that I was mostly just copying AARCH64 and not thinking too hard about what it was doing, but my intuition is that `Arch_p_arch_update_eq` is a strict extension of `Arch_p_asid_table_update_eq` and so it's fine to do the sublocale as the first thing.